File: php/src/io/mxStylesheetCodec.php
<?php
/**
 * $Id: mxStylesheetCodec.php,v 1.10 2011-06-13 08:18:42 gaudenz Exp $
 * Copyright (c) 2006-2010, Gaudenz Alder
 */
class mxStylesheetCodec extends mxObjectCodec
{

	/**
	 * Class: mxStylesheetCodec
	 *
	 * Codec for <mxStylesheets>. This class is created and registered
	 * dynamically at load time and used implicitely via <mxCodec>
	 * and the <mxCodecRegistry>.
	 * 
	 * Constructor: mxObjectCodec
	 *
	 * Constructs a new codec for the specified template object.
	 * The variables in the optional exclude array are ignored by
	 * the codec. Variables in the optional idrefs array are
	 * turned into references in the XML. The optional mapping
	 * may be used to map from variable names to XML attributes.
	 *
	 * Parameters:
	 *
	 * template - Prototypical instance of the object to be
	 * encoded/decoded.
	 * exclude - Optional array of fieldnames to be ignored.
	 * idrefs - Optional array of fieldnames to be converted to/from
	 * references.
	 * mapping - Optional mapping from field- to attributenames.
	 */
	function mxStylesheetCodec($template)
	{
		parent::mxObjectCodec($template);
	}
	
	/**
	 * Override <mxObjectCodec.encode>.
	 */
	function encode($enc, $obj)
	{
		$node = $enc->document->createElement($this->getName());
		
		foreach ($obj->styles as $i => $value)
		{
			$styleNode = $enc->document->createElement("add");
			
			if (isset($i))
			{
				$styleNode->setAttribute("as", $i);
				
				foreach ($style as $j => $value)
				{
					$value = $this->getStringValue($j, $value);
					
					if (isset($value))
					{
						$entry = $enc->document->createElement("add");
						$entry->setAttribute("value", $value);
						$entry->setAttribute("as", $j);
						$styleNode->appendChild($entry);
					}
				}
				
				if ($styleNode->getChildCount() > 0)
				{
					$node->appendChild($styleNode);
				}
			}
		}
		
	    return node;
	}
		
	/**
	 * Returns the string for encoding the given value.
	 */
	function getStringValue($key, $value)
	{
		return (!function_exists($value) && !is_object($value)) ? $value : null;
	}
	
	/**
	 * Override <mxObjectCodec.decode>.
	 */
	function decode($dec, $node, &$into = null)
	{
		$id = $node->getAttribute("id");
		$obj = (in_array($id, $dec->objects)) ? $dec->objects[$id] : null;
		
		if (!isset($obj))
		{
			if (isset($into))
			{
				$obj = $into;
			}
			else
			{
				$tmp = get_class($this->template);
				$obj = new $tmp();
			}
			
			if (isset($id))
			{
				$dec->putObject($id, $obj);
			}
		}

		$node = $node->firstChild;
		
		while (isset($node))
		{
			if (!$this->processInclude($dec, $node, $obj) &&
				$node->nodeName == "add")
			{
				$as = $node->getAttribute("as");
				
				if (strlen($as) > 0)
				{
					$extend = $node->getAttribute("extend");

					$style = (strlen($extend) > 0 &&
						isset($obj->styles[$extend])) ?
						array_slice($obj->styles[$extend], 0) :
						null;					
					
					if (!isset($style))
					{
						$style = array();
					}
					
					$entry = $node->firstChild;
					
					while (isset($entry))
					{
						if ($entry->nodeType == XML_ELEMENT_NODE)
						{
							$key = $entry->getAttribute("as");
						 	
						 	if ($entry->nodeName == "add")
						 	{
							 	$text = $entry->textContent;
							 	$value = null;
							 	
							 	if (isset($text) && strlen($text) > 0)
							 	{
							 		$value = mxUtils::evaluate($text);
							 	}
							 	else
							 	{
							 		$value = $entry->getAttribute("value");
								}
								
								if ($value != null)
								{
									$style[$key] = $value;
								}
						 	}
						 	else if ($entry->nodeName == "remove")
						 	{
						 		unset($style[$key]);
						 	}
						}
						
						$entry = $entry->nextSibling;
					}
				
					$obj->putCellStyle($as, $style);
				}
			}
			
			$node = $node->nextSibling;
		}
		
		return $obj;
	}

}

mxCodecRegistry::register(new mxStylesheetCodec(new mxStylesheet()));
?>

File: dotnet/src/io/mxStylesheetCodec.cs
// Copyright (c) 2007-2008, Gaudenz Alder
using System;
using System.Collections.Generic;
using System.Text;
using System.Xml;

namespace com.mxgraph
{
    /// <summary>
    /// Codec for mxStylesheets. This class is created and registered
    /// dynamically at load time and used implicitely via mxCodec
    /// and the mxCodecRegistry.
    /// </summary>
    public class mxStylesheetCodec : mxObjectCodec
    {

        /// <summary>
        /// Constructs a new stylesheet codec.
        /// </summary>
        public mxStylesheetCodec() : this(new mxStylesheet(), null, null, null) { }

        /// <summary>
        /// Constructs a new stylesheet codec for the given template.
        /// </summary>
        public mxStylesheetCodec(Object template) : this(template, null, null, null) { }

        /// <summary>
        /// Constructs a new stylesheet codec for the given arguments.
        /// </summary>
        public mxStylesheetCodec(Object template, String[] exclude, String[] idrefs,
                Dictionary<string, string> mapping)
            : base(template, exclude, idrefs, mapping) { }

        /// <summary>
        /// Encode the given mxStylesheet.
        /// </summary>
        public override XmlNode Encode(mxCodec enc, Object obj)
        {
            XmlElement node = enc.Document.CreateElement(GetName());
    		
		    if (obj is mxStylesheet)
		    {
			    mxStylesheet stylesheet = (mxStylesheet) obj;

                foreach (KeyValuePair<string, Dictionary<string, Object>> entry in stylesheet.Styles)
                {
                    XmlElement styleNode = enc.Document.CreateElement("add");
                    styleNode.SetAttribute("as", entry.Key);

                    foreach (KeyValuePair<string, Object> entry2 in entry.Value)
                    {
                        XmlElement entryNode = enc.Document.CreateElement("add");
					    entryNode.SetAttribute("as", entry2.Key);
                        entryNode.SetAttribute("value", getStringValue(entry2));
					    styleNode.AppendChild(entryNode);
                    }

				    if (styleNode.ChildNodes.Count > 0)
				    {
					    node.AppendChild(styleNode);
				    }
			    }
		    }
    		
		    return node;
        }

        /// <summary>
        /// Returns the string for encoding the given value.
        /// </summary>
        protected string getStringValue(KeyValuePair<string, Object> entry)
        {
            if (entry.Value is Boolean)
            {
                return ((Boolean)entry.Value) ? "1" : "0";
            }

            return entry.Value.ToString();
        }

        /// <summary>
        /// Decodes the given mxStylesheet.
        /// </summary>
        public override Object Decode(mxCodec dec, XmlNode node, Object into)
        {
            Object obj = null;

            if (node is XmlElement)
            {
                string id = ((XmlElement)node).GetAttribute("id");
                obj = (dec.Objects.ContainsKey(id)) ? dec.Objects[id] : null;

                if (obj == null)
                {
                    obj = into;

                    if (obj == null)
                    {
                        obj = CloneTemplate(node);
                    }

                    if (id != null && id.Length > 0)
                    {
                        dec.PutObject(id, obj);
                    }
                }

                node = node.FirstChild;

                while (node != null)
                {
                    if (!ProcessInclude(dec, node, obj) && node.Name.Equals("add") && node is XmlElement)
                    {
                        string name = ((XmlElement) node).GetAttribute("as");

                        if (name != null && name.Length > 0)
                        {
                            string extend = ((XmlElement) node).GetAttribute("extend");
                            Dictionary<string, Object> style = (extend != null && ((mxStylesheet)obj).Styles.ContainsKey(extend)) ?
                                    ((mxStylesheet)obj).Styles[extend] : null;

                            if (style == null)
                            {
                                style = new Dictionary<string, Object>();
                            }
                            else
                            {
                                style = new Dictionary<string, Object>(style);
                            }

                            XmlNode entry = node.FirstChild;

                            while (entry != null)
                            {
                                if (entry is XmlElement)
                                {
                                    XmlElement entryElement = (XmlElement)entry;
                                    string key = entryElement.GetAttribute("as");

                                    if (entry.Name.Equals("add"))
                                    {
                                        string text = entryElement.Value;
                                        Object value = null;

                                        if (text != null && text.Length > 0)
                                        {
                                            value = mxUtils.Eval(text);
                                        }
                                        else
                                        {
                                            value = entryElement.GetAttribute("value");
                                        }

                                        if (value != null)
                                        {
                                            style[key] = value;
                                        }
                                    }
                                    else if (entry.Name.Equals("remove"))
                                    {
                                        style.Remove(key);
                                    }
                                }

                                entry = entry.NextSibling;
                            }

                            ((mxStylesheet) obj).PutCellStyle(name, style);
                        }
                    }

                    node = node.NextSibling;
                }
            }

            return obj;
        }
    }

}
